include/pmacc/particles/algorithm/ForEach.hpp

```diff
@@ -22,157 +22,447 @@
-                            constexpr uint32_t frameSize = pmacc::math::CT::volume<SuperCellSize>::type::value;
-                            constexpr uint32_t numWorkers = T_numWorkers;
+                return pb.getFirstFrame(superCellIdx);
+            }
 
-                            uint32_t const workerIdx = cupla::threadIdx(acc).x;
+            template<typename T_ParBox>
+            static DINLINE auto next(T_ParBox& pb, typename T_ParBox::FramePtr const& framePtr)
+            {
+                return pb.getNextFrame(framePtr);
+            }
 
-                            DataSpace<dim> const superCellIdx(
-                                mapper.getSuperCellIndex(DataSpace<dim>(cupla::blockIdx(acc))));
+            template<typename T_ForEach>
+            static DINLINE auto createCtx(T_ForEach& forEachParticleInFrame)
+            {
+                return lockstep::makeVar<uint32_t>(forEachParticleInFrame, 0u);
+            }
+        };
 
-                            auto const& superCell = pb.getSuperCell(superCellIdx);
-                            uint32_t const numPartcilesInSupercell = superCell.getNumParticles();
+        //! Policy to reverse iterate over particles/frames of a supercell.
+        struct Reverse
+        {
+            template<typename T_ParBox>
+            static DINLINE auto getFirst(T_ParBox& pb, DataSpace<T_ParBox::Dim> const& superCellIdx)
+            {
+                return pb.getLastFrame(superCellIdx);
+            }
 
+            template<typename T_ParBox>
+            static DINLINE auto next(T_ParBox& pb, typename T_ParBox::FramePtr const& framePtr)
+            {
+                return pb.getPreviousFrame(framePtr);
+            }
 
-                            // end kernel if we have no particles
-                            if(numPartcilesInSupercell == 0)
-                                return;
+            template<typename T_ForEach>
+            static DINLINE auto createCtx(T_ForEach& forEachParticleInFrame)
+            {
+                return lockstep::makeVar<lcellId_t>(forEachParticleInFrame, lcellId_t(0u));
+            }
+        };
 
-                            using FramePtr = typename T_ParBox::FramePtr;
-                            FramePtr frame = pb.getFirstFrame(superCellIdx);
+        /** Execute a particle or frame functor for each particle/frame.
+         *
+         * @tparam T_AccessTag Type of the tag to mark if the particle or frame interface will be used.
+         *                     valid options: detail::CallParticleFunctor or detail::CallFrameFunctor
+         * @tparam T_Order Iteration order. valid options: Forward or Reverse
+         * @tparam T_ParBox Type of the particle box to traverse.
+         * @tparam T_numWorkers Number of lockstep workers.
+         */
+        template<typename T_AccessTag, typename T_Order, typename T_ParBox, uint32_t T_numWorkers>
+        struct ForEachParticle
```

Comment:
Matter of preference, but I would make this a class given it has a private state